dev-tools/openapi-manager/src/spec_files_generic.rs

```diff
@@ -0,0 +1,852 @@
+                let api_version = file.version();
+                let entry = self
+                    .spec_files
+                    .entry(ident.clone())
+                    .or_insert_with(ApiFiles::new)
+                    .spec_files
+                    .entry(api_version.clone());
+
+                match entry {
+                    Entry::Vacant(vacant_entry) => {
+                        vacant_entry.insert(T::make_item(file));
+                    }
+                    Entry::Occupied(mut occupied_entry) => {
+                        match occupied_entry.get_mut().try_extend(file) {
+                            Ok(()) => (),
+                            Err(error) => self.load_error(error),
+                        };
+                    }
+                };
+            }
+            Err(error) => {
+                self.load_error(error);
+            }
+        }
+    }
+
+    /// Load the "latest" symlink for a versioned API
+    ///
+    /// On failure, warnings or errors are recorded.
+    pub fn load_latest_link(
+        &mut self,
+        ident: &ApiIdent,
+        links_to: ApiSpecFileName,
+    ) {
+        let Some(api) = self.apis.api(ident) else {
+            let error =
+                anyhow!("link for unknown API {:?} ({})", ident, links_to);
+            if T::MISCONFIGURATIONS_ALLOWED {
+                self.load_warning(error);
+            } else {
+                self.load_error(error);
+            }
+
+            return;
+        };
+
+        if !api.is_versioned() {
+            let error = anyhow!(
+                "link for non-versioned API {:?} ({})",
+                ident,
+                links_to
+            );
+            if T::MISCONFIGURATIONS_ALLOWED {
+                self.load_warning(error);
+            } else {
+                self.load_error(error);
+            }
+            return;
+        }
+
+        let api_files =
+            self.spec_files.entry(ident.clone()).or_insert_with(ApiFiles::new);
+        if let Some(previous) = api_files.latest_link.replace(links_to) {
+            // unwrap(): we just put this here.
+            let new_link = api_files.latest_link.as_ref().unwrap().to_string();
+            self.load_error(anyhow!(
+                "API {:?}: multiple \"latest\" links (at least {}, {})",
+                ident,
+                previous,
+                new_link,
+            ));
+        }
+    }
+
+    /// Returns the underlying set of files loaded
+    pub fn into_map(self) -> BTreeMap<ApiIdent, ApiFiles<T>> {
+        self.spec_files
+    }
+}
+
+/// Describes a set of OpenAPI documents and associated "latest" symlink for a
+/// given API
+///
+/// Parametrized by `T` because callers use newtypes around `ApiSpecFile` to
+/// avoid confusing them.  See the documentation on [`ApiSpecFilesBuilder`].
+#[derive(Debug)]
+pub struct ApiFiles<T> {
+    spec_files: BTreeMap<semver::Version, T>,
+    latest_link: Option<ApiSpecFileName>,
```

Comment:
Could you remind me when `latest_link` is None? Does that happen only for lockstep APIs? If so would it make sense to only define `latest_link` for versioned APIs?

Reply:
This structure represents whatever was found, so if a symlink is just missing from the filesystem, this value would be `None` even for a versioned API.  And yes it would be `None` for lockstep APIs.

As you probably noticed, there's a lot of overlap between blessed files, local files, and generates files, but also some different constraints.  e.g., for the generated source, "latest" should be `Some` iff it's a versioned API.  For the blessed source, that _should_ be true, but could conceivably not be.  For local, all bets are off.  With all of these there's a tension between type safety vs. lots of repetition and extra complexity.  I was trying to strike a balance here.